Week2

mikhailgladchenko opened the pull request:
solutions for week2 tasks

mikhailgladchenko committed:
adding solutions for string1,list1,wordcount and matrix

mikhailgladchenko committed:
some changes to get_dictionary function and changes with more meaninful varibles names.

mikhailgladchenko committed:
removing unnessary file

mikhailgladchenko committed:
PEP8 inspected code

mikhailgladchenko committed:
implemented duplicate functions not using lambdas,PEP8 inspected code

mikhailgladchenko committed:
implemented _str__ ,PEP8 inspected code

mikhailgladchenko committed:
PEP8 inspected code

mikhailgladchenko committed:
implementtion with Counter class,PEP8 inspected code

mikhailgladchenko committed:
quicksort implementation,PEP8 inspected code

mikhailgladchenko committed:
quicksort implementation,PEP8 inspected code

mikhailgladchenko committed:
with changes in print_top function

mikhailgladchenko committed:
with changes to remove print function

mikhailgladchenko committed:
PEP8 inspected code

mikhailgladchenko committed:
PEP8 inspected code

mikhailgladchenko committed:
implemented match_ends_without_lambda function with list comprehension,PEP8 expected code

mikhailgladchenko committed:
implemented isvalid method to check the validity of matrices for multiplication,PEP8 expected code

mikhailgladchenko committed:
solution for task1 from week2

mikhailgladchenko committed:
solution for task2 from week2

mikhailgladchenko committed:
solution for task4 from week2

mikhailgladchenko committed:
solutions for task3 week2

mikhailgladchenko committed:
just testing week2 file addition to git

mikhailgladchenko committed:
Delete just_test.py

mikhailgladchenko committed:
final solution for task4 of week2

mikhailgladchenko committed:
solution for ex4 of week2

mikhailgladchenko committed:
solution for ex3 of week2

frizzby commented on week2/ex4_done.py:
```diff
@@ -0,0 +1,42 @@
+    print(users)
+    sorted_users_by_age = sorted(users)
+    print(sorted_users_by_age)
+
```
```python
print(User("mikhail", 57) == User("mikhail", 57))
```
Check this test case.

frizzby replied:
what about this? Did you check it?

frizzby commented on week2/ex1_done.py:
```diff
@@ -0,0 +1,31 @@
+    return new_url[0:-1]
+
+
+def main():
```
Please add more test cases. Test corner cases (like empty string, etc) as well.

frizzby commented on week2/ex2_done.py:
```diff
@@ -0,0 +1,30 @@
+    return new_url[0:-1]
+
+
+def main():
```
Please add more test cases.

frizzby commented on week2/ex3_done.py:
```diff
@@ -0,0 +1,104 @@
+    """
+
+    def __init__(self, values=None, *, mutator=_identity):
+        self._mutator = mutator
```
What is `mutator`? Why you need it here?

frizzby commented on week2/ex3_done.py:
```diff
@@ -0,0 +1,104 @@
+    def __init__(self, values=None, *, mutator=_identity):
+        self._mutator = mutator
+        if values is None:
+            self._counter = collections.Counter()
```
The tasks asks to implement your own Counter class. Not to wrap around the standard one. 
Your code should have no references to `collections.Counter`

mikhailgladchenko committed:
final solution for ex4 of week2

mikhailgladchenko committed:
final solution for ex3 of week2

mikhailgladchenko committed:
my own counter class for ex3 of week2

mikhailgladchenko committed:
enhanced solution for ex1 of week2

mikhailgladchenko committed:
enhanced solution for ex2 of week2

mikhailgladchenko committed:
final solution for ex4 of week2

frizzby commented on week2/ex3_mycounter_done.py:
```diff
@@ -0,0 +1,39 @@
+            if counter < n:
+                mc[key] = value
+            counter += 1
+        return mc
```
This has not the same return value as in [`collections.Counter`](https://docs.python.org/3/library/collections.html#collections.Counter.most_common). Please make it equivalent.

frizzby commented on week2/ex3_mycounter_done.py:
```diff
@@ -0,0 +1,39 @@
+        counter = 0
+        mc = {}
+        for key, value in sorted(self.mydict.items(), key=lambda kv: kv[1], reverse=True):
+            if counter < n:
```
Instead of counting n items, you can just slice n first elements: `sorted(...)[:n]`

frizzby commented on week2/ex3_done.py:
```diff
@@ -0,0 +1,98 @@
+# 3.	Elements counting. It refers to the sections 1.12 and 1.4-1.5 of the Python Cookbook. Implement the Counter class. It should only support operations like __iter__, __getitem__ and .most_common(n). Try to make it as efficient as possible in terms of performance. Think of an underlying data structure that should be used and also a base class.
```
Please remove this file.

frizzby commented on week2/ex2_done.py:
```diff
@@ -0,0 +1,35 @@
+
+
+def main():
+    url = "page=2&foo=bar&x=y&utm_source=ss&utm_content=ccc"
```
This code has a few flaws.
Add more test cases and you'll see it.

This code does not retain the original params order.

frizzby commented on week2/ex1_done.py:
```diff
@@ -0,0 +1,39 @@
+# 1.	Write a function that will parse a url query string like "page=2&foo=bar&x=y&utm_source=ss&utm_content=ccc" and return a query string where all the params are sorted by their names like this "foo=bar&page=2&x=y" and all the "utm_" parameters are removed.
```
_all the "utm\_" parameters are removed_ here means the all parameters that START with `utm_`. But your code removes all that contain _utm\__ substring.

frizzby commented on week2/ex1_done.py:
```diff
@@ -0,0 +1,39 @@
+
+    for key in keys_sorted:
+
+        new_url += "{}={}{}".format(key, params_dict[key], "&")
```
Don't use `+=` or `+` to build strings as it is very inefficient. Strings in python are immutable, and each concatenation creates a new string which is slow and takes memory. And also it hurts readability. 

Please use [`str.join()`](https://docs.python.org/3/library/stdtypes.html#str.join) instead.

frizzby commented on week2/ex1_done.py:
```diff
@@ -0,0 +1,39 @@
+    for entry in string_splitted:
+        key = value = ""
+        try:
+            key = entry.split("=")[0]
```
`key, value = entry.split()`

frizzby commented on week2/ex1_done.py:
```diff
@@ -0,0 +1,39 @@
+def parse_url(url):
+    string_splitted = url.split("&")
+    params_dict = {}
+    for entry in string_splitted:
```
What you are doing here is basically filtering a list. This is very common operation in Python and it is done best with [list comprehensions](https://docs.python.org/3.6/tutorial/datastructures.html#list-comprehensions):

```python
filtered = [x for x in string_splitted if not x.startswith("utm_")]
```

Another approach is to use [`filter()`](http://book.pythontips.com/en/latest/map_filter.html):
```python
filtered = filter(lambda x: not x.startswith("utm_"), string_splitted)
```


The basic operations on lists are: `filter()`, `map()`, `reduce()` ([read](http://book.pythontips.com/en/latest/map_filter.html)). Keep them in mind when dealing with lists. You can solve almost any task just with this three functions.

mikhailgladchenko committed:
with corrections for test case print(User("mikhail", 57) == User("mikhail", 57))

mikhailgladchenko committed:
using list comprehension for filtering and join

mikhailgladchenko committed:
using list comprehension for filtering and join

mikhailgladchenko committed:
changed the implementation of most_common function

mikhailgladchenko committed:
different most_common function implementations

mikhailgladchenko commented:
Yan,
I’ve just pushed new code for ex3_mycounter_done with different implementations of most_common function. I have there the one that slices the list (most_common) and another one that uses heapq.nlargest api(most_common_heap).I’m not sure that the last one works as it should.Can we take a look togather on Monday?
Regards,Mikhail

From: Yan Davidovich [mailto:notifications@github.com]
Sent: Wednesday, January 3, 2018 3:11 PM
To: frizzby/coyote-leg <coyote-leg@noreply.github.com>
Cc: Mikhail Gladchenko <Mikhail_Gladchenko@epam.com>; Author <author@noreply.github.com>
Subject: Re: [frizzby/coyote-leg] Week2 (#3)


@frizzby requested changes on this pull request.

________________________________

In week2/ex3_mycounter_done.py<https://github.com/frizzby/coyote-leg/pull/3#discussion_r159546881>:

> +            for value in values:

+                self.mydict[value] += 1

+        else:

+            self.mydict = {}

+

+    def __repr__(self):

+        return repr(self.mydict)

+

+    def most_common(self, n):

+        counter = 0

+        mc = {}

+        for key, value in sorted(self.mydict.items(), key=lambda kv: kv[1], reverse=True):

+            if counter < n:

+                mc[key] = value

+            counter += 1

+        return mc

This has not the same return value as in collections.Counter<https://docs.python.org/3/library/collections.html#collections.Counter.most_common>. Please make it equivalent.

________________________________

In week2/ex3_mycounter_done.py<https://github.com/frizzby/coyote-leg/pull/3#discussion_r159547634>:

> +

+        if values is not None:

+            self.mydict = dict.fromkeys(values, 0)

+            for value in values:

+                self.mydict[value] += 1

+        else:

+            self.mydict = {}

+

+    def __repr__(self):

+        return repr(self.mydict)

+

+    def most_common(self, n):

+        counter = 0

+        mc = {}

+        for key, value in sorted(self.mydict.items(), key=lambda kv: kv[1], reverse=True):

+            if counter < n:

Instead of counting n items, you can just slice n first elements: sorted(...)[:n]

________________________________

In week2/ex3_done.py<https://github.com/frizzby/coyote-leg/pull/3#discussion_r159547714>:

> @@ -0,0 +1,98 @@

+# 3.   Elements counting. It refers to the sections 1.12 and 1.4-1.5 of the Python Cookbook. Implement the Counter class. It should only support operations like __iter__, __getitem__ and .most_common(n). Try to make it as efficient as possible in terms of performance. Think of an underlying data structure that should be used and also a base class.

Please remove this file.

________________________________

In week2/ex2_done.py<https://github.com/frizzby/coyote-leg/pull/3#discussion_r159548467>:

> +        except IndexError:

+            pass

+        if "utm_" not in key:

+            params_dict[key] = value

+    keys_original = (params_dict.keys())

+    new_url = ""

+

+    for key in keys_original:

+

+        new_url += "{}={}{}".format(key, params_dict[key], "&")

+

+    return None if new_url == str("=&") else new_url[0:-1]

+

+

+def main():

+    url = "page=2&foo=bar&x=y&utm_source=ss&utm_content=ccc"

This code has a few flaws.
Add more test cases and you'll see it.

This code does not retains the original params order.

________________________________

In week2/ex1_done.py<https://github.com/frizzby/coyote-leg/pull/3#discussion_r159548783>:

> @@ -0,0 +1,39 @@

+# 1.   Write a function that will parse a url query string like "page=2&foo=bar&x=y&utm_source=ss&utm_content=ccc" and return a query string where all the params are sorted by their names like this "foo=bar&page=2&x=y" and all the "utm_" parameters are removed.

all the "utm_" parameters are removed here means the all parameters that START with utm_. But your code removes all that contain utm_ substring.

________________________________

In week2/ex1_done.py<https://github.com/frizzby/coyote-leg/pull/3#discussion_r159549154>:

> +    params_dict = {}

+    for entry in string_splitted:

+        key = value = ""

+        try:

+            key = entry.split("=")[0]

+            value = entry.split("=")[1]

+        except IndexError:

+            pass

+        if "utm_" not in key:

+            params_dict[key] = value

+    keys_sorted = sorted(params_dict.keys())

+    new_url = ""

+

+    for key in keys_sorted:

+

+        new_url += "{}={}{}".format(key, params_dict[key], "&")

Don't use += or + to build strings as it is very inefficient. Strings in python a immutable, and each concatenation creates new string. And also it hurts readability.

Please use ["".join()](https://docs.python.org/3/library/stdtypes.html#str.join) instead.

________________________________

In week2/ex1_done.py<https://github.com/frizzby/coyote-leg/pull/3#discussion_r159549386>:

> @@ -0,0 +1,39 @@

+# 1.   Write a function that will parse a url query string like "page=2&foo=bar&x=y&utm_source=ss&utm_content=ccc" and return a query string where all the params are sorted by their names like this "foo=bar&page=2&x=y" and all the "utm_" parameters are removed.

+

+

+def parse_url(url):

+    string_splitted = url.split("&")

+    params_dict = {}

+    for entry in string_splitted:

+        key = value = ""

+        try:

+            key = entry.split("=")[0]

key, value = entry.split()

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/frizzby/coyote-leg/pull/3#pullrequestreview-86514315>, or mute the thread<https://github.com/notifications/unsubscribe-auth/Agzo8RpcvJxOlN201cPkNXwlp-TUJ5q8ks5tHAkUgaJpZM4RKKEA>.

mikhailgladchenko committed:
fixed the bug with most_common_heap function

mikhailgladchenko committed:
added unit testing

mikhailgladchenko committed:
changes with removed dictionary

mikhailgladchenko committed:
changes with removed dictionary and unit tests added

mikhailgladchenko committed:
added new unit test case

mikhailgladchenko committed:
added new unit test case

mikhailgladchenko committed:
unit tests for ex1 in saparate file

mikhailgladchenko committed:
unit tests for ex2 in saparate file
